main.c

```diff
@@ -81,7 +90,22 @@ unsigned char write_buffer[4096];
 
 void apply_pin_state(const pio_spi_inst_t *spi, bool state) {
     pio_spi_enable_outputs(spi->pio, spi->sm, state, PIN_SCK, PIN_MOSI, PIN_MISO);
-    gpio_set_dir(PIN_CS, state? GPIO_OUT : GPIO_IN);
+    gpio_set_dir(cs_pins[active_cs_pin], state? GPIO_OUT : GPIO_IN);
+    current_pin_state = state;
+}
+
+static void set_cs_pin(uint8_t cs) {
+    // make sure new cs is initalized
+    gpio_init(cs_pins[cs]);
+    gpio_put(cs_pins[cs], 1); // assumes cs is disabled on switch
```

Comment:
No need to _assume_: `set_cs_pin` is currently only called while handling `S_CMD_S_SPI_CS` and CS is only asserted while handling `S_CMD_O_SPIOP`.

I think if you apply the initialization (see below) to all pins, this `gpio_put` won't even be necessary.